src/Workspaces/Core/Portable/Indentation/AbstractIndentationService.AbstractIndenter.cs

```diff
@@ -108,8 +115,36 @@ public IndentationResult GetDesiredIndentation(FormattingOptions.IndentStyle ind
                     token, previousNonWhitespaceOrPreprocessorLine, lastNonWhitespacePosition);
             }
 
-            protected abstract IndentationResult GetDesiredIndentationWorker(
-                SyntaxToken token, TextLine previousLine, int lastNonWhitespacePosition);
+            public bool TryGetSmartTokenIndentation(out IndentationResult indentationResult)
+            {
+                indentationResult = default;
+                if (!ShouldUseTokenIndenter(out var token))
+                {
+                    return false;
+                }
+
+                // var root = document.GetSyntaxRootSynchronously(cancellationToken);
+                var sourceText = Tree.GetText(CancellationToken);
+
+                var formatter = CreateSmartTokenFormatter();
+                var changes = formatter.FormatTokenAsync(Document.Project.Solution.Workspace, token, CancellationToken)
```

Comment:
smart token formatter actually need to change buffer? this probably only work for certain case but not any corner cases..

Reply:
also, why not making this method async?

Reply:
hmmm. or at least explain why this must be synchronous when it is not bound to UI thread.

Reply:
You're right.  It doesn't need to be synchronous.  I'll likely update to be async in a future PR.  Note: it's a bit wonky since the main editor API this has to go through is sync.  But we should at least be honest and make this all async, and push all the blocking to ISmarIndent.